prov/rxm/src/rxm_ep.c

```diff
@@ -887,10 +1029,23 @@ rxm_ep_send_common(struct rxm_ep *rxm_ep, const struct iovec *iov, void **desc,
 	assert(count <= rxm_ep->rxm_info->tx_attr->iov_limit);
 
 	ret = ofi_cmap_get_handle(rxm_ep->util_ep.cmap, dest_addr, &handle);
-	if (OFI_UNLIKELY(ret))
+	if (OFI_UNLIKELY(ret)) {
+		if (OFI_UNLIKELY(ret != -FI_EAGAIN))
+			return ret;
+		fastlock_acquire(&rxm_ep->util_ep.cmap->lock);
+		rxm_conn = container_of(handle, struct rxm_conn, handle);
+		ret = rxm_ep_postponed_send(rxm_ep, rxm_conn, context, count, iov, desc,
+					     data_len, data, flags, tag, comp_flags,
+					     (data_len <= rxm_ep->rxm_info->tx_attr->inject_size ?
+					      pool : &rxm_ep->buf_pools[RXM_BUF_POOL_TX_LMT]),
+					     op);
+		fastlock_release(&rxm_ep->util_ep.cmap->lock);
 		return ret;
+	}
```

Comment:
If it's possible to continue with TX processing (setting up tx_entry, copying data to tx_buf, etc), we should do that and finally enqueue the resulting tx_entry to postponed list. This would enable us to pipeline the tx processing while the connection establishment is pending. This could improve the latency for initial messages.

Reply:
Unfortunately, we fi_inject doesn't allocate TX entry. to avoid any branches in critical path (when connection is established), I've decided to move this in the separate function

Reply:
It'd be nice if we can make the decision of avoiding the branch based on backing performance data. Otherwise there would be code duplication and increased maintenance cost.

Reply:
I'm in favor of your idea to reduce code duplication. Currently proposed solution need to be refactored, because it looks complicated and can simplified. I'm looking for the best way to do this.